R/ntfy.R

```diff
@@ -150,71 +168,95 @@ trim_slash <- function(s) {
 #' @seealso \url{https://ntfy.sh/docs/subscribe/api/#json-message-format}
 #'
 #' @export
-ntfy_history <- function(since = "all",
-                         topic = ntfy_topic(),
-                         server = ntfy_server(),
-                         auth = NULL,
+ntfy_history <- function(since    = "all",
+                         topic    = ntfy_topic(),
+                         server   = ntfy_server(),
+                         auth     = ntfy_auth(),
+                         username = ntfy_username(),
+                         password = ntfy_password(),
                          ...) {
-  qry <- list(poll = 1, since = since, ...)
-  resp <- httr::GET(url = paste(server, topic, "json", sep = "/"), query = qry, config = auth)
-  resp <- httr::content(resp, "text")
-  resp <- gsub("\\n", "DBL_NEWLINE", resp, fixed = TRUE)
-  resp <- strsplit(resp, "\\n")[[1]]
-  res <- Reduce(rbind, lapply(resp, unjson))
-  non_missing_cols <- apply(res, 2, function(x) any(!is.na(x)))
-  res[, non_missing_cols]
-}
-
-#' unJSON edited string
-#' @keywords internal
-unjson <- function(x) {
-  x <- gsub("DBL_NEWLINE", "\\n", x, fixed = TRUE)
-  y <- list2DF(jsonlite::fromJSON(x))
-  for (col in c("id", "time", "event", "topic", "title", "message",
-                "priority", "tags", "click", "actions", "attachment")) {
-    if (!utils::hasName(y, col)) {
-      y[[col]] <- NA
-    }
+  query <- list(
+    poll = 1,
+    since = since,
+    ...
+  )
+  
+  resp <- httr2::request(server) |> 
+    httr2::req_url_path_append(topic) |> 
+    httr2::req_url_path_append("json") |> 
+    httr2::req_url_query(!!!query) |> 
+    httr2::req_method("GET") |> 
+    httr2::req_user_agent("ntfy (https://github.com/jonocarroll/ntfy)") |> 
+    req_add_auth_if_needed(auth, username, password) |>
+    httr2::req_perform()
+  
+  if (httr2::resp_has_body(resp)) {
+    # ntfy returns NDJSON (newline delimited), which has to be handled with
+    # jsonlite::stream_in(), which requires it to be a connection object
+    res <- resp |> 
+      httr2::resp_body_raw() |> 
+      rawConnection() |> 
+      jsonlite::stream_in(simplifyDataFrame = TRUE, verbose = FALSE) |> 
+      as.data.frame()
```

Comment:
```suggestion
    con <- resp |> 
      httr2::resp_body_raw() |> 
      rawConnection()
    on.exit(close(con))
    res <- 
      jsonlite::stream_in(con, simplifyDataFrame = TRUE, verbose = FALSE) |> 
      as.data.frame()
```
This connection isn't closed, and leads to some warnings 

```
Warning message:
closing unused connection 4 (httr2::resp_body_raw(resp)) 
```

Reply:
Oh man, *that's* where it's coming from! I kept trying to hunt down where the unclosed connection was and couldn't find it. Fixed now.